File: frontend/cnr/src/feature/profile/domain/repository/ProfileRepository.ts
import { ErrorResponse } from "../../../../services/model/commun";
import {
  ChildResponse,
  InstitutionResponse,
  PhaseResponse,
  ProfileResponse,
  UsersResponse,
} from "../../data/dtos/ProfileDtos";

export interface ProfileRepository {
  GetProfile(
    token: string,
    permission: string
  ): Promise<ProfileResponse | ErrorResponse>;

  GetUsers(
    token: string,
    permission: string
  ): Promise<UsersResponse | ErrorResponse>;

  GetAllUsers(
    token: string,
    permission: string
  ): Promise<UsersResponse | ErrorResponse>;

  GetInstituations(
    token: string,
    permission: string
  ): Promise<InstitutionResponse | ErrorResponse>;
  GetChildOfInstitutions(
    id: string,
    name: string,
    token: string,
    permission: string
  ): Promise<ChildResponse | ErrorResponse>;
  GetCurrentPhase(
    token: string,
    permission: string
  ): Promise<PhaseResponse | ErrorResponse>;

  AddPk(
    token: string,
    permission: string,
    pk: string
  ): Promise<boolean | ErrorResponse>;

  UpdateFirstLastName(
    token: string,
    permission: string,
    firstName: string,
    lastName: string,
    avatar: string | undefined
  ): Promise<boolean | ErrorResponse>;

  UpdatePassword(
    token: string,
    permission: string,
    oldPassword: string,
    newPassword: string
  ): Promise<boolean | ErrorResponse>;

  SendOTP(token: string, email: string): Promise<boolean | ErrorResponse>;

  ConfirmOTP(token: string, otp: string): Promise<boolean | ErrorResponse>;

  VerifySignature(
    token: string,
    permission: string,
    signature: string,
    randomValue: string
  ): Promise<boolean | ErrorResponse>;

  UpdateUser(
    token: string,
    permission: string,
    userId: string,
    newType: string,
    status?: boolean
  ): Promise<boolean | ErrorResponse>;
}

File: backend/feature/home/profile/repository/profile_repository.go
type ProfileRepository interface {
	GetProfile(userId string) (*User, error)
	GetFolders(folder *fabric.FolderMetadata) (*fabric.FolderMetadata, error)
	GetCurrentPhase() (*Phase, error)
	AddPK(userId string, pk string) error
	UpdateFirstLastName(userId string, firstName string, lastName string, avatar string) error
	UpdatePassword(userId string, oldPassword string, newPassword string) error
	VerifyDigitalSignature(userId string, signature string, randomValue string) error
	GetAllUsers() ([]*User, error)
	UpdateUserType(userId string, newType string) error
	UpdateUserStatus(userId string, status bool) error
}

func (p *profileRepository) UpdateUserType(userId string, newType string) error {
	// Update user type in the database
	query := `UPDATE users SET type = $1 WHERE id = $2`
	_, err := p.db.Exec(query, newType, userId)
	if err != nil {
		return err
	}
	return nil
}

func (p *profileRepository) UpdateUserStatus(userId string, status bool) error {
	// Update user status in the database
	query := `UPDATE users SET status = $1 WHERE id = $2`
	_, err := p.db.Exec(query, status, userId)
	if err != nil {
		return err
	}
	return nil
} 